### Teorema de Bayes

- Distribució Conjunta
- Distribució Condicional - Versemblança -
- Distribució a priori
- Distribució a posteriori

$$
P\left(Y|X\right) = \frac{P\left(Y\right)\,P\left(X|Y\right)}{P\left(X\right)}
$$

Nomes la part de dalt &rarr; Posteriori (si normalitzem tenim la probabilitat) proporcional priori * likelihood

- Això és el teorema de Bayes.
- Lu interessant és: Si x i y són dues variables que tenen una certa relació (x variable independent i y dependent). Són variables estocastiques. cada una d'elles li podem asociar una distribució. Bayes diu, que la distribució de la variable dependent, marginal (per ella mateixa) distribució a priori (P(Y)) té relació directa donat una valor de la variable dependent (distribució a posteriori). 
- X és les mostres dades, una matriu (sepal_lenght, sepal_width) i Y és la classe (versicolor, virginica, ...)
- P(Y) si tenim una distribució a priori, podem saber la distribució a posterior un cop la variable dependent. 
- Relaciona distribució condicional amb una marginal. 
- La relació és la part de la dreta P(X|Y) / P(X).
- Si tenim una flor nova, podem saber la probabilitat de que sigui de cada una de les classes.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
pd.set_option('display.max.columns', None)
import seaborn as sns
sns.set_theme()

In [2]:
iris = sns.load_dataset("iris")
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [3]:
from scipy.stats import gaussian_kde

#### Variable independent: X

- X és un vector amb els valors dels atributs (columnes): $\vec{X}=\left(x_1, x_2, \dots, x_m\right)$
- per exemple, la observació 0 és el vector: $\vec{X_0}=\left(5.1, 3.5, 1.4, 0.2\right)$ 

#### Variable dependent: Y

- en el cas genèric seria també un vector: $\vec{Y}=\left(y_1, y_2, \dots, y_l\right)$
- en problemes de classificació acostuma a ser una categoria
- en aquest cas: $Y\in\{\text{"setosa", "virginica", "versicolor"}\}$, per exemple: $Y_0=\text{"setosa"}$

### Joint input distribution: $P\left(X\right)$

- és la distribució de la variable independent, un vector amb tots els atributs que tenim en el dataset
- en aquest cas serà una distribució conjunta de 4 dimensions
- la podem estimar a partir de l'histograma conjunt o taula de contingencies o amb un KDE si la variable és contínua

In [4]:
joint_pdf = gaussian_kde(iris[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']].transpose())

In [5]:
# Li passem les dades:
iris[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']].head(2)

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2

- podem avaluar la densitat per cada un dels exemples

In [6]:
example = iris.iloc[np.random.randint(150)]
example

sepal_length       4.3
sepal_width        3.0
petal_length       1.1
petal_width        0.1
species         setosa
Name: 13, dtype: object

In [7]:
species, values = example.species, tuple(example[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']])
jointpdf = joint_pdf.evaluate(values).item()
print('Example is %s ... density: %6.4f' %(example.species, jointpdf))

Example is setosa ... density: 0.2713


- però de què ens serveix això ??? ... 

Per:
- Fins ara hem fet servir totes les observacions P(X).
- Sabem que això és P(X) &rarr; variable independent (els números).
- Tenim una distribucio conjunta de variables.
- De moment no hi ha relacio entre els numeros i la categoria.
- Mirem a la montanyeta, quina densitat li toca.
- Amb les observacions hem contruit això.
- Amb les observacions, mirem quina densitat ha quedat.
- Com que algú ha posat etiquetes, que passa si fem la distribució de les x, però nomès d'una classe en concret.
- Ara el fem nomès les versicolors, no totes. Ho farem per setosa, virginica (totes):
- pdf(x) on nomès x tal que la seva y és versicolor &rarr; X|Y=versicolor
- P(X|Y) és aquesta, només amb una y fixe.

### Conditional distribution: $P\left(X|Y\right)$

In [8]:
setosa = iris.loc[iris.species == 'setosa'][['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]

In [9]:
pdf_setosa = gaussian_kde(setosa.transpose(), bw_method = 'scott')

In [10]:
virginica = iris.loc[iris.species == 'virginica'][['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]

In [11]:
pdf_virginica = gaussian_kde(virginica.transpose(), bw_method = 'scott')

In [12]:
versicolor = iris.loc[iris.species == 'versicolor'][['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]

In [13]:
pdf_versicolor = gaussian_kde(versicolor.transpose(), bw_method = 'scott')

### Likelihood function (versemblança)  $\mathcal{L}\left(\theta, X\right)\approxeq\; P\left(X|Y\right)$

- quan parlem de distribució condicional, també es pot dir de varsemblança
- perque podem evaluar a totes les funcions de vers i veure com de versemblant és que aquell conjunt sigui d'aquella classe.
- knn era en base distancies i aqui en base densitats. Els dos son models/algorismes de classificació.

In [14]:
idx = np.random.randint(50)
example = iris.loc[iris.species == 'setosa'].reset_index().iloc[idx]
example

index               19
sepal_length       5.1
sepal_width        3.8
petal_length       1.5
petal_width        0.3
species         setosa
Name: 19, dtype: object

In [15]:
pdf_setosa.evaluate(tuple(example[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]))

array([4.87548216])

In [16]:
pdf_virginica.evaluate(tuple(example[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]))

array([9.59332501e-85])

In [17]:
pdf_versicolor.evaluate(tuple(example[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']]))

array([2.62279246e-53])

In [18]:
example = iris.iloc[np.random.randint(150)]
species, values = example.species, tuple(example[['sepal_length', 'sepal_width', 'petal_length', 'petal_width']])
pdfSetosa = pdf_setosa.evaluate(values).item()
pdfVirginica = pdf_virginica.evaluate(values).item()
pdfVersicolor = pdf_versicolor.evaluate(values).item()
print('Example is %s ... likelihood: setosa %6.4f, virginica %6.4f, versicolor %6.4f' %(example.species, pdfSetosa, pdfVirginica, pdfVersicolor))

Example is setosa ... likelihood: setosa 7.9622, virginica 0.0000, versicolor 0.0000


- però això són densitats, no són probabilitats !!!!
- es l'alçada en aquell punt de la pdf

- Una cosa es dir, hem construit la pdf condicionada a setosa, saps que qualsevol cosa serà una setosa. Estem assumint la hipotesis.
- L'altre és evaluar la probabilitat que una cosa sigui una setosa donats uns numeros.
- P(Y) és 33% probabilitat que sigui setosa/vir/vers (sense saber res més) &rarr; és la priori
- La formula s'ha de passar de densitat a probabilitat i normalitzar (dividir) per P(X) &rarr; ens ho podem petar perque sempre serà igual i la part de dalt serà proporcional.
- Mostra represesntativa peruqe sino no representa la realitat. Suficients per cobrir tot el domini.

### Prior distribution: $P\left(Y\right)$

- és la distribució de la variable dependent $Y$
- si no en tenim informació l'estimem a partir de la mostra (igual que la de la variable dependent $X$)
- però podria ser que en tinguessim informació i si fos així la fariem servir (exemple eleccions)

In [19]:
iris.species.value_counts(normalize = True)

setosa        0.333333
versicolor    0.333333
virginica     0.333333
Name: species, dtype: float64

In [20]:
prior = iris.species.value_counts(normalize = True)

### Posterior distribution:  $P\left(Y|X\right)$

- és la distribució condicional de la variable dependent $Y$ donat un valor de la variable independent $X$
- es relaciona amb la distribució condicional $P\left(X|Y\right)$ segons el *Teorema de Bayes*
- el **Teorema de Bayes** quantifica com varia la distribució de la variable dependent 𝑌 deguda al fet d'haver observat una certa evidència (un determinat valor de la variable independent 𝑋)

### <center>Bayes Theorem</center>

$$
\text{if}\;X,\,Y\;\text{dependent variables}
\\
\text{ }
\\
P\left(X, Y\right) = P\left(X|Y\right)\,P\left(Y\right) = P\left(Y|X\right)\,P\left(X\right)
\\
\text{ }
\\
\text{then}
\\
\text{ }
\\
P\left(Y|X\right) = \frac{P\left(Y\right)\,P\left(X|Y\right)}{P\left(X\right)}
$$

### Exemple

- donada una observació d'una Iris qualsevol, **i sense cap altra tipus d'informació adicional**, és a dir sense haver mesurat ni pètals ni sèpals, la nostra "creença" sobre a quína espècie pertany és la que ens dona la *distribució a priori* $P\left(Y\right)$, en aquest cas, d'un 33.3% de probabilitat per cada espècie.
- però si mesurem els seus pètals i sèpals i observem els valors:
    - sepal length: 5.2
    - sepal width: 3.3
    - petal length: 1.5
    - petal width: 0.3
- aleshores, la probabilitat de cada espècie és:

$$
\\
P\left(Y=\text{'setosa'}\,|\,X=\left(5.2, 3.3, 1.5, 0.3\right)\right)=\frac{P\left(Y=\text{'setosa'}\right)\,P\left(X=\left(5.2, 3.3, 1.5, 0.3\right)\,|\,Y=\text{'setosa'}\right)}{P\left(X=\left(5.2, 3.3, 1.5, 0.3\right)\right)}
\\
\text{ }
\\
P\left(Y=\text{'virginica'}\,|\,X=\left(5.2, 3.3, 1.5, 0.3\right)\right)=\frac{P\left(Y=\text{'virginica'}\right)\,P\left(X=\left(5.2, 3.3, 1.5, 0.3\right)\,|\,Y=\text{'virginica'}\right)}{P\left(X=\left(5.2, 3.3, 1.5, 0.3\right)\right)}
\\
\text{ }
\\
P\left(Y=\text{'versicolor'}\,|\,X=\left(5.2, 3.3, 1.5, 0.3\right)\right)=\frac{P\left(Y=\text{'versicolor'}\right)\,P\left(X=\left(5.2, 3.3, 1.5, 0.3\right)\,|\,Y=\text{'versicolor'}\right)}{P\left(X=\left(5.2, 3.3, 1.5, 0.3\right)\right)}
$$

- que substituint per les distribucions que hem estimat de la mostra, dona:

$$
\\
P\left(Y=\text{'setosa'}\,|\,X=\left(5.2, 3.3, 1.5, 0.3\right)\right)=\frac{0.33\;\text{pdf_setosa}\left(5.2, 3.3, 1.5, 0.3\right)}{\text{joint_pdf}\left(5.2, 3.3, 1.5, 0.3\right)}
\\
\text{ }
\\
P\left(Y=\text{'virginica'}\,|\,X=\left(5.2, 3.3, 1.5, 0.3\right)\right)=\frac{0.33\;\text{pdf_virginica}\left(5.2, 3.3, 1.5, 0.3\right)}{\text{joint_pdf}\left(5.2, 3.3, 1.5, 0.3\right)}
\\
\text{ }
\\
P\left(Y=\text{'versicolor'}\,|\,X=\left(5.2, 3.3, 1.5, 0.3\right)\right)=\frac{0.33\;\text{pdf_versicolor}\left(5.2, 3.3, 1.5, 0.3\right)}{\text{joint_pdf}\left(5.2, 3.3, 1.5, 0.3\right)}
$$

### Inferència

- en aquestes expressions, el denominador $P\left(X\right)$ és sempre comú a totes (és només un factor d'escala), per tant podem obviar-lo i considerar:

$$
P\left(Y|X\right) \propto P\left(Y\right)\,P\left(X|Y\right)
$$

- aleshores, podem fer inferència sobre $Y$ mirant per quína classe la probabilitat a posteriori és màxima:

$$
\arg\max_{c\in Y}{P\left(c|X\right) = P\left(c\right)\;P\left(X|c\right)}
$$

- per la mostra que tenim de Iris les tres classes són "a priori" igual de probables (en la realitat no tindria perquè ser així), o sigui:

$$
\forall{c\in Y},\;P\left(c\right) = 0.33
$$

- en general, **sempre que les classes siguin equiprobables**, la probabilitat a posteriori és directament proporcional a la versemblança de les dades:

$$
\forall{c}\in Y,\; P\left(Y=c|X\right) \propto P\left(X|Y=c\right)
$$

- per les dades observades $\vec{X}=\left(5.2, 3.3, 1.5, 0.3\right)$ la classe més versemblant (o més probable) serà:

In [21]:
values = (5.2, 3.3, 1.5, 0.3)
pdfSetosa = pdf_setosa.evaluate(values).item()
pdfVirginica = pdf_virginica.evaluate(values).item()
pdfVersicolor = pdf_versicolor.evaluate(values).item()
print('likelihood: setosa %6.4f, virginica %6.4f, versicolor %6.4f' %(pdfSetosa, pdfVirginica, pdfVersicolor))

likelihood: setosa 3.9084, virginica 0.0000, versicolor 0.0000


- les probabilitats les podem obtenir normalitzant les versemblances obtingudes (perquè les probabilitats de totes les classes han de sumar 1),

In [22]:
sumpdf = pdfSetosa +pdfVirginica +pdfVersicolor
print('posterior probability: setosa %6.4f, virginica %6.4f, versicolor %6.4f' %(pdfSetosa/sumpdf, pdfVirginica/sumpdf, pdfVersicolor/sumpdf))

posterior probability: setosa 1.0000, virginica 0.0000, versicolor 0.0000


### Valors de la bbdd

In [23]:
exemple = iris.iloc[np.random.randint(150)]
species, values = exemple.species, exemple[[0,1,2,3]].to_numpy().astype('float').T
species, values

('virginica', array([6.3, 3.4, 5.6, 2.4]))

In [24]:
exemple = iris.iloc[np.random.randint(150)]
species, values = exemple.species, exemple[[0,1,2,3]].to_numpy().astype('float').T

pdfSetosa = pdf_setosa.evaluate(values).item()
pdfVirginica = pdf_virginica.evaluate(values).item()
pdfVersicolor = pdf_versicolor.evaluate(values).item()
print('likelihood: setosa %6.4f, virginica %6.4f, versicolor %6.4f, species %s' %(pdfSetosa, pdfVirginica, pdfVersicolor, species))

likelihood: setosa 0.0000, virginica 0.0002, versicolor 2.1581, species versicolor


In [25]:
sumpdf = pdfSetosa +pdfVirginica +pdfVersicolor
print('posterior probability: setosa %6.4f, virginica %6.4f, versicolor %6.4f' %(pdfSetosa/sumpdf, pdfVirginica/sumpdf, pdfVersicolor/sumpdf))
print(sumpdf)

posterior probability: setosa 0.0000, virginica 0.0001, versicolor 0.9999
2.1583266409550905


El P(Y|X) és una densitat. Per passar-ho a probabilitat. 

In [26]:
jointpdf = joint_pdf.evaluate(values).item()
print('posterior probability: setosa %6.4f, virginica %6.4f, versicolor %6.4f' %(pdfSetosa/jointpdf, pdfVirginica/jointpdf, pdfVersicolor/jointpdf))
print(jointpdf)

posterior probability: setosa 0.0000, virginica 0.0008, versicolor 9.0245
0.23914087165520867


<hr/>

El classificador bayesia es molt potent però dificil d'aplicar, hem de fer el naive (reduit notebook 46).

<br/>

És dificil d'aplicar perquè necessitem moltes dades per omplir l'espai i no quedin buits, pensa que estem agrupant les dades amb locs per tant hem de tenir moltes de moltes per tenir totes les possibles combinacions possibles. Per això és millor estudiar les variables per separat. 